### PR TITLE
Ephemeral event support for appservices

### PR DESCRIPTION
**Pull request by Half-Shot**
I couldn't find many places to stick an unstable, and to be honest this implementation was fairly trivial to write.

Implements https://github.com/matrix-org/matrix-doc/pull/2409

### CODE REVIEW

File: src/appservice/Appservice.ts [hunk 1]
```diff
@@ -612,6 +649,15 @@ export class Appservice extends EventEmitter {
                 }
             }
 
+            if (this.registration["de.sorunome.msc2409.push_ephemeral"] && req.body["de.sorunome.msc2409.ephemeral"]) {
+                for (let event of req.body["de.sorunome.msc2409.ephemeral"]) {
+                    LogService.info("Appservice", `Processing ephemeral event of type ${event["type"]}`);
+                    event = await this.processEphemeralEvent(event);
+                    // These events aren't tied to rooms, so just emit them as `ephemeral_event`
+                    this.emit("ephemeral_event", event);
```
Comment (Half-Shot) [inline src/appservice/Appservice.ts:657]: This could potentially be `unstable_ephemeral_event`, WDYT?

Comment (turt2live) [reply src/appservice/Appservice.ts:657]: meh, enabling the flag using some non-standard namespace and bolded as **Experimental** is probably fine enough.

File: src/appservice/Appservice.ts [hunk 2]
```diff
@@ -508,6 +517,23 @@ export class Appservice extends EventEmitter {
         }
     }
 
+    /**
+     * Adds a preprocessor to the event pipeline. When this appservice encounters an ephemeral event, it
+     * will try to run it through the preprocessors it can in the order they were added.
+     * @param {IPreprocessor} preprocessor the preprocessor to add
+     */
+    public addEphemeralEventPreprocessor(preprocessor: IPreprocessor): void {
```
Comment (turt2live) [inline src/appservice/Appservice.ts:525]: In practice I think we can support the pre-existing pre-processors: people have to opt in to this.

**Review by turt2live — CHANGES_REQUESTED** [review]
woo, thanks! Overall it seems fine (and surprisingly small) - just a small thing about the event preprocessor stuff.

If you get a chance, tests would be nice 😇

**Comment by turt2live** [comment]
(will address during merge)

**Pull request merged**